core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1541,7 +1541,8 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
       initial: O2
   ): Stream[F2, Signal[F2, O2]] =
     Stream.eval(SignallingRef.of[F2, O2](initial)).flatMap { sig =>
-      Stream(sig).concurrently(evalMap(sig.set))
+      def setAll(ch: Chunk[O]): Pull[F2, Nothing, Unit] = Pull.eval(ch.traverse_(sig.set))
+      Stream(sig).concurrently(underlying.unconsFlatMap(setAll).streamNoScope)
```

Comment:
I'd understand if we replaced `evalMap(sig.set)` with `foreach(sig.set)`.

Otherwise, aren't we essentially re-implementing `foreach` here?

Reply:
No. If you look at the code of `foreach`, you can see that it wraps each execution of the effect function `f` with a `Pull.eval`, whereas here we are using a `traverse_` on the entire chunk.

Reply:
So why can't we make this optimization to `foreach` ?

Reply:
Oh, you mean why not do the `traverse_` directly in the body of `foreach`? I have tried a few times, but it does not work. There are some problems with the interruption scope depending on the actions.


However this code, where we only do a `sig.set`, should not have that problem.